common/src/main/java/com/wynntils/wc/utils/scoreboard/quests/QuestManager.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.wc.utils.scoreboard.quests;
+
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.mc.event.WynntilsScoreboardUpdateEvent;
+import com.wynntils.wc.utils.scoreboard.ScoreboardLine;
+import com.wynntils.wc.utils.scoreboard.ScoreboardManager;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import net.minecraft.ChatFormatting;
+import net.minecraft.server.ServerScoreboard;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import org.apache.commons.lang3.StringUtils;
+
+public class QuestManager {
+
+    private static final String TRACKED_QUEST_STRING = "§6§lTracked Quest:";
+
+    private static QuestInfo currentQuest = null;
+
+    @SubscribeEvent
+    public static void onScoreboardUpdate(WynntilsScoreboardUpdateEvent event) {
+        if (!event.getChangeMap().containsKey(WynntilsScoreboardUpdateEvent.ChangeType.Quest)) return;
+
+        Set<WynntilsScoreboardUpdateEvent.Change> changes =
+                event.getChangeMap().get(WynntilsScoreboardUpdateEvent.ChangeType.Quest);
+
+        for (WynntilsScoreboardUpdateEvent.Change change : changes) {
+            if (change.method() == ServerScoreboard.Method.REMOVE) {
+                if (TRACKED_QUEST_STRING.equals(change.line())) {
+                    questTrackingStopped();
+                }
+            } else { // Method is CHANGE
+                // Use copy constructor to prevent threading issues
+                List<ScoreboardLine> reconstructedScoreboard =
+                        new ArrayList<>(ScoreboardManager.getReconstructedScoreboard());
+
+                List<String> questLines = reconstructedScoreboard.stream()
+                        .dropWhile(scoreboardLine -> !scoreboardLine.line().equals(TRACKED_QUEST_STRING))
+                        .takeWhile(scoreboardLine -> !scoreboardLine.line().matches("À+"))
+                        .map(scoreboardLine -> ChatFormatting.stripFormatting(scoreboardLine.line()))
+                        .filter(Objects::nonNull)
+                        .skip(1)
+                        .toList();
+
+                if (questLines.isEmpty()) {
+                    WynntilsMod.error("QuestManager: questLines was empty.");
+                }
+
+                currentQuest = new QuestInfo(
+                        questLines.get(0),
+                        StringUtils.joinWith(" ", questLines.stream().skip(1).collect(Collectors.toList())));
+
+                // Only update currentQuest once per batch.
```

Comment:
Can you get `changes` that is like (`CHANGE`: ...., `REMOVE`: ...)? If so, could you not end up in a situation where you setup a new currentQuest here, and then replace it with `null` right after?

Or, hm, you are just looking if the header is removed. And reasonably the header could not be removed and the actual quest lines are changed. I guess...

TL;DR: This seems like it could be working, but it's more of the kind "no obvious problems" instead of "obviously no problems".

Reply:
Order should be correct as it is the order of the packets. And the header is only removed if the quest tracking stops. I know it feels a bit strange, but this should work fine.